python/ray/train/_internal/dataset_spec.py

```diff
@@ -179,6 +179,7 @@ def get_dataset_shards(
                     ).repeat()
                     # In windowed mode, we re-apply the preprocessor on each iteration.
                     if self.preprocessor:
+                        # TODO: Replace with self.preprocessor.transform when possible.
```

Comment:
Does this mean that we'll need to support a `batch_format="zero_copy"` or similar and do the dispatch within the transform UDF, since we may not be able to determine the dataset format of the pipeline?

Reply:
I think we can pluck it as a workaround, right? I'll follow this up in another PR

Reply:
I think plucking should work, since neither `.window()` nor `.repeat()` should change the dataset format.

Reply:
We should be able to do the right thing automatically by checking the dataset format.